File: app/src/main/java/com/arash/altafi/lazyload/java/adapter/AdapterCelebrities.java
package com.arash.altafi.lazyload.java.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.arash.altafi.lazyload.R;
import com.arash.altafi.lazyload.java.model.ResponseCelebritiesJava;
import com.arash.altafi.lazyload.java.utils.OnLoadMoreListener;
import com.bumptech.glide.Glide;

import java.util.List;

public class AdapterCelebrities extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    Context context;
    List<ResponseCelebritiesJava> list;
    private OnLoadMoreListener onLoadMoreListener;
    private int sizeIndex;
    private int lastVisible;
    private final int countLoadMore = 2;
    private boolean isLoading = false;
    private final int VIEW_TYPE_RESPONSE = 1;
    private final int VIEW_TYPE_LOADMORE = 2;

    public AdapterCelebrities(Context context, RecyclerView recyclerView, List<ResponseCelebritiesJava> list) {
        this.context = context;
        this.list = list;

        LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (layoutManager != null) {
                    sizeIndex = layoutManager.getItemCount();
                }
                if (layoutManager != null) {
                    lastVisible = layoutManager.findLastVisibleItemPosition();
                }

                if (!isLoading && sizeIndex <= (lastVisible + countLoadMore)) {
                    if (onLoadMoreListener != null) {
                        onLoadMoreListener.loadMore();
                    }
                    isLoading = true;
                }

            }
        });
    }

    public void setLoading(boolean loading) {
        isLoading = loading;
    }

    public void setOnLoadMoreListener(OnLoadMoreListener onLoadMoreListener) {
        this.onLoadMoreListener = onLoadMoreListener;
    }

    @Override
    public int getItemViewType(int position) {
        if (list.get(position) != null)
            return VIEW_TYPE_RESPONSE;
        else
            return VIEW_TYPE_LOADMORE;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_RESPONSE) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_bazigar, parent, false);
            return new JavaViewHolder(view);
        } else if (viewType == VIEW_TYPE_LOADMORE) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_load_more, parent, false);
            return new JavaLoadMoreViewHolder(view);
        }

        return null;

    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof JavaViewHolder) {
            ResponseCelebritiesJava celebrity = list.get(position);
            Glide.with(context).load(celebrity.getImage()).into(((JavaViewHolder) holder).imageView);
        } else if (holder instanceof JavaLoadMoreViewHolder) {
            ((JavaLoadMoreViewHolder) holder).progressBar.setIndeterminate(true);
        }

    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public static class JavaViewHolder extends RecyclerView.ViewHolder {
        ImageView imageView;

        public JavaViewHolder(@NonNull View itemView) {
            super(itemView);
            imageView = itemView.findViewById(R.id.ivCelebrity);
        }
    }

    public static class JavaLoadMoreViewHolder extends RecyclerView.ViewHolder {

        ProgressBar progressBar;

        public JavaLoadMoreViewHolder(@NonNull View itemView) {
            super(itemView);
            progressBar = itemView.findViewById(R.id.pr_Loading_More);
        }
    }
}

File: app/src/main/java/com/arash/altafi/lazyload/kotlin/adapter/AdapterCelebrities.kt
package com.arash.altafi.lazyload.kotlin.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.arash.altafi.lazyload.R
import com.arash.altafi.lazyload.kotlin.model.ResponseCelebrities
import com.arash.altafi.lazyload.kotlin.utils.OnLoadMoreListenerKotlin
import com.bumptech.glide.Glide

class AdapterCelebrities(list2: ArrayList<ResponseCelebrities?>, recyclerView: RecyclerView) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var list: ArrayList<ResponseCelebrities?> = ArrayList()
    private var onLoadMoreListener: OnLoadMoreListenerKotlin? = null
    private var sizeIndex = 0
    private var lastVisible = 0
    private val countLoadMore = 2
    private var isLoading = false

    init {

        this.list = list2

        val layoutManager = recyclerView.layoutManager as LinearLayoutManager?
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                sizeIndex = layoutManager!!.itemCount
                lastVisible = layoutManager.findLastVisibleItemPosition()
                if (!isLoading && sizeIndex <= lastVisible + countLoadMore) {
                    if (onLoadMoreListener != null) {
                        onLoadMoreListener!!.loadMore()
                    }
                    isLoading = true
                }
            }
        })
    }

    fun setLoading(loading: Boolean) {
        isLoading = loading
    }

    fun setOnLoadMoreListener(onLoadMoreListener: OnLoadMoreListenerKotlin?) {
        this.onLoadMoreListener = onLoadMoreListener
    }

    override fun getItemViewType(position: Int): Int {
        return if (list[position] != null) VIEW_TYPE_RESPONSE else VIEW_TYPE_LOADMORE
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == VIEW_TYPE_RESPONSE) {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.layout_bazigar, parent, false)
            return KotlinViewHolder(view)
        }
        else if (viewType == VIEW_TYPE_LOADMORE) {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.item_load_more, parent, false)
            return KotlinLoadMoreViewHolder(view)
        }
        return null!!
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is KotlinViewHolder) {
            val celebrity = list[position]
            Glide.with(holder.itemView.context).load(celebrity!!.image).into(holder.imageView)
        }
        else if (holder is KotlinLoadMoreViewHolder) {
            (holder).progressBar.isIndeterminate = true
        }
    }

    override fun getItemCount(): Int = list.size

    companion object {
        private const val VIEW_TYPE_RESPONSE = 1
        private const val VIEW_TYPE_LOADMORE = 2
    }

}

class KotlinViewHolder(itemView : View) : RecyclerView.ViewHolder(itemView) {

    val imageView : ImageView = itemView.findViewById(R.id.image_celebrity)

}

class KotlinLoadMoreViewHolder(itemView : View) : RecyclerView.ViewHolder(itemView) {

    val progressBar : ProgressBar = itemView.findViewById(R.id.pr_Loading_More)

}